MapleServer2/Types/FieldNavigator.cs

```diff
@@ -0,0 +1,244 @@
+    public Position FindPositionFromCoordS(CoordS coordF)
+    {
+        return Mesh.positionNear3DPoint(coordF.X, coordF.Y, coordF.Z, horizontalRange: 15, verticalRange: 15);
+    }
+
+    /// <summary>
+    /// Check if the given position is valid.
+    /// </summary>
+    public bool PositionIsValid(Position position)
+    {
+        return Mesh.positionIsValid(position);
+    }
+
+    public Position? FindClosestUnobstructedPosition(Shape shape, Position position, int radius)
+    {
+        if (!PositionIsValid(position))
+        {
+            return null;
+        }
+
+        Position randomPosition = Mesh.generateRandomPositionLocally(position, radius);
+        if (!PositionIsValid(randomPosition))
+        {
+            return null;
+        }
+
+        Position unobstructedPosition = Mesh.findClosestUnobstructedPosition(shape, CollisionContext, randomPosition, radius);
+        if (!PositionIsValid(unobstructedPosition))
+        {
+            return null;
+        }
+
+        return unobstructedPosition;
+    }
+
+    public CoordS? FindClosestUnobstructedCoordS(Shape shape, CoordS coordS, int radius)
+    {
+        Position position = FindPositionFromCoordS(coordS);
+        return FindClosestUnobstructedCoordS(shape, position, radius);
+    }
+
+    public CoordS? FindClosestUnobstructedCoordS(Shape shape, Position position, int radius)
+    {
+        if (!PositionIsValid(position))
+        {
+            return null;
+        }
+
+        Position randomPosition = Mesh.generateRandomPositionLocally(position, radius);
+        if (!PositionIsValid(randomPosition))
+        {
+            return null;
+        }
+
+        Position unobstructedPosition = Mesh.findClosestUnobstructedPosition(shape, CollisionContext, randomPosition, radius);
+        if (!PositionIsValid(unobstructedPosition))
+        {
+            return null;
+        }
+
+        return CoordS.From(unobstructedPosition.X, unobstructedPosition.Y, Mesh.heightAtPosition(unobstructedPosition));
+    }
+
+    private List<CoordS> PathToCoordS(Path path)
+    {
+        if (path is null)
+        {
+            return null;
+        }
+
+        List<CoordS> coordS = new();
+        for (int i = 0; i < path.size(); i++)
+        {
+            Position position = path.position(i);
+            if (position.Cell < 0)
+            {
+                Console.WriteLine("Invalid position: " + position);
+                continue;
+            }
+
+            coordS.Add(CoordS.From(position.X, position.Y, Mesh.heightAtPosition(position)));
+        }
+
+        return coordS;
+    }
+
+    public void Dispose()
+    {
+        Mesh?.Dispose();
+        GC.SuppressFinalize(this);
+    }
+}
+
+public class ErrorHandler : IErrorHandler
+{
+    public override ErrorResult handle(ErrorType type, string description, IDictionary<string, string> attributes)
+    {
+        Console.WriteLine("Type: " + type + " Description: " + description + " Attributes: " + attributes);
```

Comment:
you can't really print a dictionary like this. I provide a basic error handler in the library now.

```
new PrintErrorHandler(Console.Out);
```